[5.4] Update ContenthistoryHelper to use absolute image and href paths

Attila-SWE opened the pull request:
My first PR so sorry if I messed something up. Please give me feedback then. Some considerations is that this only fixes images in images/ folder and does not fix other relative URL:s or such. Perhaps a broader fix is necessary?

Pull Request for Issue #44635.

### Summary of Changes
Check to see if values in content history includes images with relative /images url

### Testing Instructions
Open preview of content history with images in content with relative urls like images/test.png

### Actual result BEFORE applying this Pull Request
Images broken because urls are relative to /administrator/ so for example DOMAIN/administrator/images/test.png

### Expected result AFTER applying this Pull Request
Images shown and url of images is DOMAIN/images/test.png

### Link to documentations
No documentation changes for docs.joomla.org needed
No documentation changes for manual.joomla.org needed

Attila-SWE committed:
Update ContenthistoryHelper to use absolute image paths

My first PR and regarding issue https://github.com/joomla/joomla-cms/issues/44635

Attila-SWE commented:
Done! So if I understand correctly then first create issue, then create pull request referring to issue and close issue? Is there any "documentation" for this regarding Joomla that I could read up on?

fgsw commented:
If you create a PR no Issue is needed.

I don't know about Documentation for this workflow.

alikon commented on administrator/components/com_contenthistory/src/Helper/ContenthistoryHelper.php:
```diff
@@ -19,6 +19,7 @@
 use Joomla\Filesystem\File;
 use Joomla\Filesystem\Folder;
 use Joomla\Filesystem\Path;
+use JURI;
```
use
use Joomla\CMS\Uri\Uri;

alikon commented on administrator/components/com_contenthistory/src/Helper/ContenthistoryHelper.php:
```diff
@@ -79,6 +80,10 @@ public static function decodeFields($jsonString)
                 }
+
+                if (!empty($value) && is_string($value) && strpos($value, 'src="images/') !== false) {
+		            $object->$name = str_replace('src="images/', 'src="' . JURI::root() . 'images/', $value);
```
use 
 URI:root()

Attila-SWE replied:
Of course! Sorry for old school habit :-)

Attila-SWE committed:
Updated PR with no deprecated classes...

fgsw commented:
@alikon Makes it sense to test the Pull Request as ```continuous-integration/drone/pr — Build is failing```?

alikon commented:
@fgsw drone is failing at Code style step

alikon commented on administrator/components/com_contenthistory/src/Helper/ContenthistoryHelper.php:
```diff
@@ -19,6 +19,7 @@
 use Joomla\Filesystem\File;
 use Joomla\Filesystem\Folder;
 use Joomla\Filesystem\Path;
+use Joomla\CMS\Uri\Uri;
```
please alpha sort

Attila-SWE committed:
Changed order use-list

Sorry for many small changes. First PR.

alikon commented on administrator/components/com_contenthistory/src/Helper/ContenthistoryHelper.php:
```diff
@@ -79,6 +80,10 @@ public static function decodeFields($jsonString)
                     $object->$name = $subObject;
                 }
+
+                if (!empty($value) && is_string($value) && strpos($value, 'src="images/') !== false) {
```
```suggestion
                if (!empty($value) && \is_string($value) && strpos($value, 'src="images/') !== false) {
```

alikon reviewed (dismissed):
yet another code style
sorry i didn't check before

Attila-SWE commented:
NP! Thanks for helping me learn

Attila-SWE committed:
Change to global \is_string() instead of is_string()

alikon committed:
code style

hoping is the last 1 for code syle

Attila-SWE committed:
Merge pull request #1 from alikon/patch-18

code style for  your #44645

fgsw commented:
I have tested this item :white_check_mark: successfully on ecd0668530393f1171db04c054f346ae939dd059<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44645">issues.joomla.org/tracker/joomla-cms/44645</a>.</sub>

chmst commented:
Thank you very much @Attila-SWE and congratulations to your first PR!

alikon commented:
successfully tested by @Quy  on https://github.com/joomla/joomla-cms/commit/ecd0668530393f1171db04c054f346ae939dd059

alikon commented:
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44645">issues.joomla.org/tracker/joomla-cms/44645</a>.</sub>

Fedik commented:
Back to Pending status.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/44645">issues.joomla.org/tracker/joomla-cms/44645</a>.</sub>

Fedik commented:
I am sorry, the changes will not going to work reliable. But thanks for your try.
You is assuming that all images are from `images/` folder, however it not always true.

A solid fix going to be much more complex. It need something similar to what we have in SEF plugin https://github.com/joomla/joomla-cms/blob/9800738a8f457a60dff8d1d44254913e62603b1a/plugins/system/sef/src/Extension/Sef.php#L229-L259

Attila-SWE committed:
Universal absolute path

This time it checks if not absolute path for src and adds it if necessary. Not just if images/

Attila-SWE commented:
New suggested solution.

Fedik commented:
Thanks. That looks much better. Still few things:
- Check that the string actualy contain `src=` before running reg expression (as it done in SEF plugin). Can use [str_contains()](https://www.php.net/manual/uk/function.str-contains.php)
- You checking only `https`, but we should check any protocol (as it done in SEF plugin)
- Code style: https://ci.joomla.org/joomla/joomla-cms/81227/1/7

Attila-SWE commented:
Should we also fix all relative 'href' or just 'src'?

Fedik commented:
If you like you can do for `href` also, sure.

Attila-SWE committed:
Adaptation to suggestions

Added check for href
Check if string contains attribute before replacement

Attila-SWE commented:
@Fedik I would appreciate your input on new suggested change.

Attila-SWE committed:
Merge branch '5.3-dev' into patch-1

Fedik commented:
Looks better, thanks.
Fix the code style then https://ci.joomla.org/joomla/joomla-cms/83497/1/7 ,  it can be tested.

Attila-SWE committed:
Merge branch '5.3-dev' into patch-1

Attila-SWE committed:
CS-fix ContenthistoryHelper.php

Last try I hope :-)

Attila-SWE committed:
CS-fix

argh... :-)

Attila-SWE committed:
no_whitespace_in_blank_line

Attila-SWE commented:
> Looks better, thanks. Fix the code style then https://ci.joomla.org/joomla/joomla-cms/83497/1/7 , it can be tested.

Finished my n00b-wrestling with code style 🙈

Attila-SWE committed:
Merge branch '5.3-dev' into patch-1

Attila-SWE commented:
Why is drone failing? Can it be restarted?

richard67 committed:
Merge branch '5.4-dev' into patch-1
